Mathlib/RingTheory/Bialgebra.lean

```diff
@@ -0,0 +1,140 @@
+/-
+Copyright (c) 2024 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey, Kevin Buzzard
+-/
+import Mathlib.RingTheory.Coalgebra
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Bialgebras
+
+In this file we define `Bialgebra`, and provide instances for:
+
+* Commutative semirings: `CommSemiring.toBialgebra`
+
+## References
+
+* <https://en.wikipedia.org/wiki/Bialgebra>
+-/
+
+suppress_compilation
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A bialgebra over a commutative (semi)ring `R` is both an algebra and a coalgebra over `R`, such
+that the counit and comultiplication are algebra morphisms. -/
+class Bialgebra (R : Type u) (A : Type v) [CommSemiring R] [Semiring A] extends
+    Algebra R A, Coalgebra R A where
+  -- The counit is an algebra morphism
+  /-- The counit on a bialgebra preserves 1. -/
+  counit_one : counit 1 = 1
+  /-- The counit on a bialgebra preserves multiplication. Note that this is written
+  in a rather obscure way: it says that two curried maps `A →ₗ[R] A →ₗ[R]` are equal.
```

Comment:
Could there please be an explanation of why this is written in a rather obscure way? :-)

Reply:
There's an explanation on this github page but you're right: I've added it to the module docstring as an implementation note.